### DISCUSSION TITLE
Use named command variants in ConTeXt

### DISCUSSION THREAD

**Discussion by juh2** [jgm/pandoc#9604] [Q&A] [answered]
The power of ConTeXt is the ability to derive named command variants like this:

```
\defineexternalfigure
[text]
[width=\textwidth]

\defineexternalfigure
[margin]
[width=\rightmarginwidth]

\defineexternalfigure
[full]
[width=\dimexpr\textwidth+\rightmargintotal\relax]
```

Three figures are defined from the default externalfigure: one for the textarea, one for the margin and one that spans the whole textarea with the margin.

The definitions are used like this inside a \placefigure environment:

```
\externalfigure
[image.png]
[full]
[height=3cm]
```

I guess that I can pass the name "full" to the output with a filter so that I can write in Markdown something like this:

```
:::full
![ALT](image.jpg)
:::
```

Any idea how to achieve that? Has anyone yet implemented such a filter?

TIA
juh

**Comment by jgm** [comment, marked as answer]
Currently the writer has:
```hs
  return $ braces $ "\\externalfigure" <> brackets (literal src') <> dims <> clas
```
where `clas` is the first class, in brackets, if it exists, otherwise nothing, and `dims` are dimension specifiers e.g. `height=3in` in brackets, if specified, otherwise nothing.

This suggests you could just add a class, and it looks like that works:
```
% pandoc -t context
![ALT](image.jpg){.full}
\startplacefigure[title={ALT}]
{\externalfigure[image.jpg][full]}
\stopplacefigure
```

**Reply by juh2** [reply]
Wow, it is so easy!  :smile:

**Reply by massifrg** [reply]
```sh
pandoc -t context
![ALT](image.jpg){.full height=3cm}
\startplacefigure[title={ALT}]
{\externalfigure[image.jpg][height=3cm][full]}
\stopplacefigure
```

The [syntax](https://wiki.contextgarden.net/Command/externalfigure) of `\externalfigure` is:

```context
\externalfigure[FILE][NAME][OPTIONS from \setupexternalfigure]
```

so the 2nd and 3rd argument should be swapped, like this:

```context
\externalfigure[image.jpg][full][height=3cm]
```

Maybe the writer should be like this:

```haskell
return $ braces $ "\\externalfigure" <> brackets (literal src') <> clas <> dims
```

Since the 2nd and 3rd arguments of `\externalfigure` are both optional, no problem arises unless you specify *both* class and dimensions.

**Reply by jgm** [reply]
Thanks, I've made that change.
